File: example/src/main/java/me/devnatan/inventoryframework/runtime/view/Failing.java
package me.devnatan.inventoryframework.runtime.view;

import me.devnatan.inventoryframework.View;
import me.devnatan.inventoryframework.ViewConfigBuilder;
import me.devnatan.inventoryframework.context.RenderContext;
import me.devnatan.inventoryframework.runtime.ExampleUtil;
import me.devnatan.inventoryframework.state.MutableState;
import org.bukkit.Material;
import org.jetbrains.annotations.NotNull;

public class Failing extends View {

	MutableState<Integer> state = mutableState(0);

	@Override
	public void onInit(@NotNull ViewConfigBuilder config) {
		config.size(1);
		config.cancelOnClick();
		config.title("Failing Inventory");
		config.layout("  R   C  ");
	}

	@Override
	public void onFirstRender(@NotNull RenderContext render) {
		render.layoutSlot('R')
			.onRender((ctx) -> {
				if (state.get(ctx) == 0) {
					ctx.setItem(ExampleUtil.displayItem(Material.DIAMOND, "Click me to fail"));
				} else {
					throw new IllegalStateException("This item cannot be rendered");
				}
			}).onClick((ctx) -> {
				state.set(1, ctx);
				ctx.update();
			});

		render.layoutSlot('C', ExampleUtil.displayItem(Material.STONE, "Click me and I will fail"))
			.onClick((ctx) -> {
				throw new IllegalStateException("This is a failing inventory");
			});
	}
}

File: examples/minestom/src/main/kotlin/me/devnatan/inventoryframework/runtime/view/Failing.kt
package me.devnatan.inventoryframework.runtime.view

import me.devnatan.inventoryframework.View
import me.devnatan.inventoryframework.ViewConfigBuilder
import me.devnatan.inventoryframework.context.RenderContext
import me.devnatan.inventoryframework.context.SlotClickContext
import me.devnatan.inventoryframework.context.SlotRenderContext
import me.devnatan.inventoryframework.runtime.ExampleUtil.displayItem
import me.devnatan.inventoryframework.state.MutableState
import net.minestom.server.item.Material

class Failing : View() {
    var state: MutableState<Int> = mutableState(0)

    override fun onInit(config: ViewConfigBuilder) {
        config.size(1)
        config.cancelOnClick()
        config.title("Failing Inventory")
        config.layout("  R   C  ")
    }

    override fun onFirstRender(render: RenderContext) {
        render.layoutSlot('R')
            .onRender { ctx: SlotRenderContext ->
                if (state[ctx] == 0) {
                    ctx.item =
                        displayItem(
                            Material.DIAMOND,
                            "Click me to fail",
                        )
                } else {
                    throw IllegalStateException("This item cannot be rendered")
                }
            }
            .onClick { ctx: SlotClickContext ->
                state[1] = ctx
                ctx.update()
            }

        render.layoutSlot('C', displayItem(Material.STONE, "Click me and I will fail"))
            .onClick { _ ->
                throw IllegalStateException("This is a failing inventory")
            }
    }
}
